### PR TITLE
Add CBMC proof for Shadow_MatchTopic.

### PR DESCRIPTION
**Pull request by leegeth**
*Description of changes:*
Add CBMC proof for Shadow_MatchTopic API.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/shadow.c [hunk 1]
```diff
@@ -148,8 +148,8 @@ static ShadowStatus_t validateMatchTopicParameters( const char * pTopic,
  * @param[in]  thingNameLength Length of Thing Name string pointed to by pThingName.
  * @param[in]  pShadowName Shadow Name string.
  * @param[in]  shadowNameLength Length of Shadow Name string pointed to by pShadowName.
- * @param[in] pTopicBuffer Pointer to buffer for returning the topic string.
- * @param[in] pOutLength Pointer to caller-supplied memory for returning the length of the topic string.
+ * @param[in]  pTopicBuffer Pointer to the topic buffer.
+ * @param[in]  pOutLength Pointer to the length of the topic created.
```
Comment (aggarw13) [inline source/shadow.c:152]: Can this be changed to an integer value instead of a pointer?

File: source/shadow.c [hunk 2]
```diff
@@ -938,6 +938,12 @@ ShadowStatus_t Shadow_MatchTopic( const char * pTopic,
                                                            NULL,
                                                            NULL );
 
+    /* Update the output parameter for Thing name length.
+     * Shadow_MatchTopicString takes a pointer to a 8 bit unsigned integer for
+     * output parameter `Thing name length`, whereas Shadow_MatchTopic takes a
```
Comment (aggarw13) [inline source/shadow.c:943]: (Minor) Does the use of ticks "`" help in any way?

File: test/cbmc/proofs/Shadow_MatchTopic/Makefile [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#
+# Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
```
Comment (aggarw13) [inline test/cbmc/proofs/Shadow_MatchTopic/Makefile:2]:
```suggestion
# Copyright (C) 2021 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
```

File: source/shadow.c (hunk 1 above)
Comment (leegeth) [reply source/shadow.c:152]: No. The validation is done for pointers in this function.

File: source/shadow.c (hunk 2 above)
Comment (leegeth) [reply source/shadow.c:943]: No. It can be changed. Thanks.

Comment (leegeth) [inline source/shadow.c:943]:
```suggestion
     * output parameter Thing name length, whereas Shadow_MatchTopic takes a
```

File: test/cbmc/proofs/Shadow_MatchTopic/Shadow_MatchTopic_harness.c [hunk 4]
```diff
@@ -0,0 +1,52 @@
+/*
+ * AWS IoT Device Shadow v1.0.2
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
```
Comment (leegeth) [inline test/cbmc/proofs/Shadow_MatchTopic/Shadow_MatchTopic_harness.c:3]:
```suggestion
 * Copyright (C) 2021 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
```

**Review by aggarw13 — DISMISSED** [review]

File: test/cbmc/proofs/Shadow_MatchTopic/Makefile [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#
+
+HARNESS_ENTRY=harness
+HARNESS_FILE=Shadow_MatchTopic_harness
+PROOF_UID = Shadow_MatchTopic
+
+# The topic length is bounded to reduce the proof run time. This length bounds
+# the time to look for the thing name in the topic string. The minimum length
+# of the topic string is the length of the SHADOW_PREFIX ("$aws/things/")
+# plus the maximum Shadow Message Type length ("/shadow/update/documents").
```
Comment (gkwicker) [inline test/cbmc/proofs/Shadow_MatchTopic/Makefile:29]: What about the name of the thing?  They must be [at least one character in length.](https://docs.aws.amazon.com/iot/latest/apireference/API_CreateThing.html)

Also, this vallue does not agree with the definition of `TOPIC_STRING_LENGTH_MAX` on line 33 (36 vs. 50).

**Review by gkwicker — CHANGES_REQUESTED** [review]
See comment

File: test/cbmc/proofs/Shadow_MatchTopic/Makefile (hunk 5 above)
Comment (leegeth) [reply test/cbmc/proofs/Shadow_MatchTopic/Makefile:29]: Thanks for the comment. Reading again the comments, I see the issue. The difference is allotted for ThingName Length.(50-37). I will update the comment to indicate that.

**Review by gkwicker — APPROVED** [review]

**Review by aggarw13 — APPROVED** [review]

**Pull request merged**